### PR TITLE
Reproducible setuptools

### PR DESCRIPTION
**Pull request by basile-henry**
###### Motivation for this change

Part of the efforts at https://r13y.com/ to get a reproducible iso_minimal

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/setuptools/default.nix [hunk 1]
```diff
@@ -27,14 +27,22 @@ let
...
       ${python.pythonForBuild.interpreter} bootstrap.py
       ${python.pythonForBuild.interpreter} setup.py sdist --formats=gztar
     '';
 
+    # Here we untar the sdist and retar it in order to control the timestamps
+    # of all the files included
     installPhase = ''
-      echo "Moving sdist..."
-      mv dist/*.tar.gz $out
+      echo "Repackaging sdist..."
```
Comment (FRidh) [inline pkgs/development/python-modules/setuptools/default.nix:42]: please do the repackaging in the `buildPhase`, and keep the `mv` in `installPhase`. It's good practice to keep the two steps in distinct phases.

**Review by FRidh — COMMENTED** [review]
Thank you! Have you checked `bootstrapped-pip` is also reproducible?

File: pkgs/development/python-modules/setuptools/default.nix (hunk 1 above)
Comment (basile-henry) [reply pkgs/development/python-modules/setuptools/default.nix:42]: Sure I can do that :+1:

**Comment by basile-henry** [comment]
@FRidh `nix-build -A python38Packages.bootstrapped-pip --check` seems to work just fine :+1: Are there more check you would like me to do?

Someone on IRC mentioned that only `gnutar` exposes the `--mtime` flag (and not `bsdtar`). Do I need to do something special on darwin?

**Comment by FRidh** [comment]
@GrahamcOfBorg build python3.pkgs.setuptools

**Comment by r-burns** [comment]
I think the tar command is fine, bootstrapped-pip and setuptools are building and passing `--check` for me on darwin :)

**Comment by basile-henry** [comment]
@r-burns I think you might need to run `nix-build -A pythonPackages.setuptools.src --check`. Otherwise it might get that from the cache (I think). It was the derivation that makes up `setuptools`'s `src` that was previously not reproducible, not `setuptools` itself (I should have specified that better in the PR description :sweat_smile: ).

**Comment by FRidh** [comment]
@basile-henry by changing the expression the derivation is changed, so there is no issue with cache using here. Anyway, I will wait for ofborg on darwin for confirmation.

**Comment by basile-henry** [comment]
> so there is no issue with cache using here

You are right for the first build. What I am unsure about is how deep `--check` goes. I was under the impression it only checks 1 derivation deep, but maybe that's not the case? If so, how does it know how deep to check?

**Comment by FRidh** [comment]
The `src` attribute is not a fixed-output derivation. `--check` does not recurse.

**Comment by basile-henry** [comment]
Building it once on darwin is indeed sufficient to check if the tar command works. Thanks for checking that :+1: 

I just wanted to check for myself if `--check` recursed or not and as you said it does not:

```
➜ cat test.nix
let pkgs = import <nixpkgs> {};

    not-deterministic = pkgs.stdenv.mkDerivation {
      name = "date";
      buildPhase = ''
        mkdir $out
        date > $out/date
      '';
      phases = [ "buildPhase" ];
    };
in
pkgs.stdenv.mkDerivation {
  name = "test";
  src = not-deterministic;
  buildPhase = ''
    mkdir $out
    cp date $out/date
  '';
  phases = [ "unpackPhase" "buildPhase" ];
}

➜ nix-build test.nix
/nix/store/sf7p5x1wr070vkg668dlndlfjhl83mbk-test

➜ nix-build test.nix --check
checking outputs of '/nix/store/53ryaaxhd4xd8bs4jh3rm3i0dbxf9fdf-test.drv'...
unpacking sources
unpacking source archive /nix/store/q3dnl180w8b7pc8lm4j8k978jr872n8p-date
source root is date
building
/nix/store/sf7p5x1wr070vkg668dlndlfjhl83mbk-test

➜ cat $(nix-build test.nix)/date
Tue Nov 17 10:17:21 UTC 2020
```

File: pkgs/development/python-modules/setuptools/44.0.nix [hunk 2]
```diff
@@ -27,14 +27,23 @@ let
...
+      ./tag-date.patch
+    ];
+
     buildPhase = ''
       ${python.pythonForBuild.interpreter} bootstrap.py
       ${python.pythonForBuild.interpreter} setup.py sdist --formats=gztar
+
+      # Here we untar the sdist and retar it in order to control the timestamps
+      # of all the files included
+      tar -xzf dist/${pname}-${version}.post0.tar.gz -C dist/
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/setuptools/44.0.nix:40]: Is post0 always the same?

Comment (basile-henry) [reply pkgs/development/python-modules/setuptools/44.0.nix:40]: Yes, the name of the release tag `post0` is deterministic now (it used to contain the date). The patch of `setup.cfg` ensures that.

It is possible that an update of the source changes that (if the build system upstream is updated), but I would assume that the `tag-date.patch` would also need to be updated then. If/when that happens this `buildPhase` should fail loudly, so I don't think hardcoding it is an issue.

If you have any suggestion to make this better, do tell! :smile:

File: pkgs/development/python-modules/setuptools/44.0.nix [hunk 3]
```diff
@@ -27,14 +27,23 @@ let
...
+      tar -czf dist/${name} -C dist/ --mtime='1970-01-01' ${pname}-${version}.post0
```
Comment (prusnak) [inline pkgs/development/python-modules/setuptools/44.0.nix:41]: Maybe this is better?

```suggestion
      tar -czf dist/${name} -C dist/ --mtime="@$SOURCE_DATE_EPOCH" ${pname}-${version}.post0
```

File: pkgs/development/python-modules/setuptools/default.nix [hunk 4]
```diff
@@ -27,14 +27,23 @@ let
...
+    ];
+
     buildPhase = ''
       ${python.pythonForBuild.interpreter} bootstrap.py
       ${python.pythonForBuild.interpreter} setup.py sdist --formats=gztar
+
+      # Here we untar the sdist and retar it in order to control the timestamps
+      # of all the files included
+      tar -xzf dist/${pname}-${version}.post0.tar.gz -C dist/
+      tar -czf dist/${name} -C dist/ --mtime='1970-01-01' ${pname}-${version}.post0
```
Comment (prusnak) [inline pkgs/development/python-modules/setuptools/default.nix:41]: Same as above:

```suggestion
      tar -czf dist/${name} -C dist/ --mtime="@$SOURCE_DATE_EPOCH"  ${pname}-${version}.post0
```

**Review by prusnak — COMMENTED** [review]
We can replace the hardcoded value of `'1970-01-01'` with `"@$SOURCE_DATE_EPOCH"`

**Review by prusnak — APPROVED** [review]

**Pull request merged**